### PR TITLE
Fulcrum protocol

### PR DESCRIPTION
**Pull request by yashasvi-ranawat**
Added [fulcrum protocol](https://electrum-cash-protocol.readthedocs.io/en/latest/index.html) for additional APIs with tests and docs.

The additional APIs lead to slowing of get_sanitised_endpoints_for function. I parallelised that with threading; that puts the first hit still at 2s.

### CODE REVIEW

**Comment by JettScythe** [comment]
Feel free to add my instance here as well: fulcrum.jettscythe.xyz

You can find a full list of instances here: https://1209k.com/bitcoin-eye/ele.php?chain=bch

**Review by merc1er — APPROVED** [review]

**Comment by merc1er** [comment]
Thank you for yet again another quality pull request @yashasvi-ranawat.

I have known about Fulcrum for a while but never actually used it. The [Fulcrum website ](https://fulcrumserver.org)does not really give it a good image 😅. However I believe this is what powers Electron Cash, so I assume it is very reliable.

Do you think it is much more reliable than any of our current providers? If so, maybe the end goal would be to use Fulcrum as a default provider for every API call? Currently I don't see any code cash token related - [it looks like CashToken is supported](https://electrum-cash-protocol.readthedocs.io/en/latest/protocol-basics.html#cashtoken-support). Maybe at some point that could be integrated and Fulcrum could be the default provider once it has proven it is extremely reliable.

File: bitcash/network/APIs/FulcrumProtocolAPI.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+from bitcash.exceptions import InvalidEndpointURLProvided
+from bitcash.network.APIs import BaseAPI
+from bitcash.network.meta import Unspent
+from bitcash.network.transaction import Transaction, TxPart
+from bitcash.cashaddress import Address
+
+
+context = ssl.create_default_context()
+FULCRUM_PROTOCOL = "1.5.0"
```
[the remark this answers is not in the document]
Comment (merc1er) [reply bitcash/network/APIs/FulcrumProtocolAPI.py:18]: My bad, it looks like the protocol and the software have different versions. The [latest protocol version seems to be 1.5.3](https://electrum-cash-protocol.readthedocs.io/en/latest/protocol-changes.html).

[the remark this answers is not in the document]
Comment (yashasvi-ranawat) [reply bitcash/network/APIs/FulcrumProtocolAPI.py:18]: Yes, the latest version is 1.5.3; But for our usecases 1.5.0 is enough! That adds cashtoken data to the txs.

**Comment by yashasvi-ranawat** [comment]
> However I believe this is what powers Electron Cash, so I assume it is very reliable.
Yes indeed. It is reliable. But I don't know if the servers are instantly synced with the blockchain.

> Currently I don't see any code cash token related - [it looks like CashToken is supported](https://electrum-cash-protocol.readthedocs.io/en/latest/protocol-basics.html#cashtoken-support).
Cashtoken is supported! The protocol version 1.5.0 has token related fields in the API, and hence I chose that to as the base protocol.

> Maybe at some point that could be integrated and Fulcrum could be the default provider once it has proven it is extremely reliable.
That would be ideal! We can remove the default APIs from other implementations while still keeping them. I like Chaingraph approach as it is very versatile, and pyBitcash can still support it.

The full adaption of Fulcrum will come with further design debates! Bitcash is built on the idea of a static kind of library that only queries on demand, but other BCH tools like DSproof and address subscription, and fulcrum tools like peer discovery, might need a deviation from that idea (and better implementation of Fulcrum to not allow problems like: https://t.me/electroncashserver/4899).

Where the implementation stands right now: with some default endpoints, Fulcrum protocol is 100% usable for pybitcash!

**Comment by merc1er** [comment]
FYI @yashasvi-ranawat it looks like these changes are not compatible with Python 3.8: https://github.com/pybitcash/bitcash/actions/runs/9184862197/job/25257867718.

It's [almost at its end of life](https://devguide.python.org/versions/) so maybe we can drop support and replace it with 3.12 at some point. I think I'll do that unless it's a super easy fix to make it compatible with 3.8.

**Pull request merged**